src/tempfile.rs

```diff
@@ -0,0 +1,167 @@
+//! let t = TempFile::new("/tmp/testfile").unwrap();
+//! let mut f = t.as_file();
+//! f.write_all(b"hello world").unwrap();
+//! f.sync_all().unwrap();
+
+use std::ffi::{CString, OsStr};
+use std::fs;
+use std::fs::File;
+use std::os::unix::ffi::{OsStrExt, OsStringExt};
+use std::os::unix::io::FromRawFd;
+use std::path::{Path, PathBuf};
+
+use libc;
+
+use crate::errno::{errno_result, Result, Error};
+
+/// Create and remove a temporary file.  The file will be maintained for the
+/// lifetime of the `TempFile` object.
+pub struct TempFile {
+    path: PathBuf,
+    file: File,
+    removed: bool,
+}
+
+
+impl TempFile {
+
+    /// Creates the TempFile
+    ///
+    /// # Arguments
+    ///
+    /// `prefix`: The path and filename where to creat the temporary file. Six
+    /// random alphanumeric characters will be added to the end of this to form
+    /// the filename.
+    pub fn new<P: AsRef<OsStr>>(prefix: P) -> Result<TempFile> {
+        let mut os_fname = prefix.as_ref().to_os_string();
+        os_fname.push("XXXXXX");
+
+        let raw_fname = CString::new(os_fname.into_vec()).unwrap().into_raw();
+
+        // Safe because I'm sure `raw_fname` is a valid CString.
+        let fd = unsafe { libc::mkstemp(raw_fname) };
+
+        // The `raw_fname` is the same as `os_fname` (which is `prefix` with X's
+        // appended). The X's are documented as "six characters" which seem to
+        // always be alphanumerics so appears safe.
+        let c_tempname = unsafe { CString::from_raw(raw_fname) };
+        let os_tempname = OsStr::from_bytes(c_tempname.as_bytes());
+
+        if fd == -1 {
+            return errno_result();
+        }
+
+        // Safe because we checked `fd != -1` above and we uniquely own the file
+        // descriptor. This `fd` will be freed etc when `File` and thus
+        // `TempFile` goes out of scope.
+        let file = unsafe { File::from_raw_fd(fd) };
+
+        Ok(TempFile {
+            path: PathBuf::from(os_tempname),
+            file,
+            removed: false,
+        })
+    }
+
+    /// Removes the temporary file.  Calling this is optional as dropping a
+    /// `TempFile` object will also remove the file.  Calling remove explicitly
+    /// allows for better error handling.
+    pub fn remove(&mut self) -> Result<()> {
+        if self.removed {
+            return Ok(());
+        }
+        fs::remove_file(&self.path).map_err(Error::from).map(|_| {
+            self.removed = true;
+        })
+        // match fs::remove_file(&self.path) {
+        //     Err(e) => Result::Err(e),
+        //     Ok(_) => {
+        //         self.removed = true;
+        //         Ok(())
+        //     }
+        // }
+    }
+
+    /// Returns the path to the tempfile if it is currently valid
+    pub fn as_path(&self) -> &Path {
+        &self.path
+    }
+
+    /// Returns a reference to the File
+    pub fn as_file(&self) -> &File {
+        &self.file
+    }
+}
+
+impl Drop for TempFile {
```

Comment:
We should also have a test for drop. I also tested this in tempdir, you can check the code in my PR: https://github.com/andreeaflorescu/vmm-sys-util/blob/7e478e5a7d4b2b5330a927b7b953780a9fd3630d/src/tempdir.rs#L127